data/vibe/data/json.d

```diff
@@ -1779,22 +1779,23 @@ struct JsonSerializer {
 	void beginWriteArrayEntry(Traits)(size_t) {}
 	void endWriteArrayEntry(Traits)(size_t) { m_compositeStack[$-1].appendArrayElement(m_current); }
 
-	void writeValue(Traits, T)(in T value)
-		if (!is(T == Json))
+	void writeValue(Traits, T)(auto ref T value)
+		if (!is(Unqual!T == Json))
 	{
-		static if (is(T == JSONValue)) {
+		alias UT = Unqual!T;
+		static if (is(UT == JSONValue)) {
 			m_current = Json(value);
-		} else static if (isJsonSerializable!T) {
+		} else static if (isJsonSerializable!UT) {
 			static if (!__traits(compiles, () @safe { return value.toJson(); } ()))
-				pragma(msg, "Non-@safe toJson/fromJson methods are deprecated - annotate "~T.stringof~".toJson() with @safe.");
+				pragma(msg, "Non-@safe toJson/fromJson methods are deprecated - annotate "~UT.stringof~".toJson() with @safe.");
 			m_current = () @trusted { return value.toJson(); } ();
-		} else static if (isSomeString!T && !is(T == string)) {
+		} else static if (isSomeString!T && !is(UT == string)) {
 			writeValue!Traits(value.to!string);
 		} else m_current = Json(value);
 	}
 
-	void writeValue(Traits, T)(Json value) if (is(T == Json)) { m_current = value; }
-	void writeValue(Traits, T)(in Json value) if (is(T == Json)) { m_current = value.clone; }
+	void writeValue(Traits, T)(auto ref T value) if (is(Unqual!T == Json)) { m_current = value; }
+	//void writeValue(Traits, T)(auto ref in T value) if (is(Unqual!T == Json)) { m_current = value.clone; }
```

Comment:
I'm not sure what to do about this as all tests passes without it.

Reply:
I think we need both, if serializing `const(Json)` is supposed to work. I'd do it with template constraints in this case instead of `Unqual`:
```
	void writeValue(Traits, T)(auto ref T value) if (is(T == Json)) { m_current = value; }
	void writeValue(Traits, T)(auto ref T value) if (!is(T == Json) && is(T : const(Json))) { m_current = value.clone; }
```
If you add this change, it would be good to also have a unit test that tests something like a `const(C)` with `class C { Json foo; }` to make sure the second case is actually hit.

Reply:
fixed